src/test/scala/edu/berkeley/cs/rise/opaque/JoinSuite.scala

```diff
@@ -0,0 +1,357 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package edu.berkeley.cs.rise.opaque
+
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.internal.SQLConf
+
+trait JoinSuite extends OpaqueSQLSuiteBase with SQLHelper {
+  import spark.implicits._
+
+  def numPartitions: Int
+
+  def queries = Seq(
+    "SELECT * FROM testData LEFT SEMI JOIN testData2 ON key = a",
+    "SELECT * FROM testData LEFT SEMI JOIN testData2",
+    "SELECT * FROM testData LEFT JOIN testData2",
+    "SELECT * FROM testData RIGHT JOIN testData2",
+    "SELECT * FROM testData LEFT JOIN testData2 WHERE key = 2",
+    "SELECT * FROM testData JOIN testData2 ON key = a",
+    "SELECT * FROM testData JOIN testData2 ON key = a and key = 2",
+    "SELECT * FROM testData JOIN testData2 ON key = a where key = 2",
+    "SELECT * FROM testData LEFT JOIN testData2 ON key = a",
+    "SELECT * FROM testData RIGHT JOIN testData2 ON key = a where key = 2",
+    "SELECT * FROM testData left JOIN testData2 ON (key * a != key + a)",
+    "SELECT * FROM testData right JOIN testData2 ON (key * a != key + a)",
+    "SELECT * FROM testData ANTI JOIN testData2 ON key = a",
+    "SELECT * FROM testData LEFT ANTI JOIN testData2",
+    "SELECT * FROM testData t1 JOIN " +
+      "testData2 t2 ON t1.key = t2.a JOIN testData3 t3 ON t2.a = t3.a",
+    "SELECT * FROM testData t1 JOIN " +
+      "testData2 t2 ON t1.key = t2.a JOIN " +
+      "testData3 t3 ON t2.a = t3.a JOIN " +
+      "testData t4 ON t1.key = t4.key"
+  )
+  def failingQueries = Seq()
+  def unsupportedQueries = Seq(
+    "SELECT * FROM testData JOIN testData2",
+    "SELECT * FROM testData JOIN testData2 WHERE key = 2",
+    "SELECT * FROM testData RIGHT JOIN testData2 WHERE key = 2",
+    "SELECT * FROM testData JOIN testData2 WHERE key > a",
+    "SELECT * FROM testData FULL OUTER JOIN testData2 WHERE key = 2",
+    "SELECT * FROM testData FULL OUTER JOIN testData2",
+    "SELECT * FROM testData FULL OUTER JOIN testData2 WHERE key > a",
+    "SELECT * FROM testData full JOIN testData2 ON (key * a != key + a)"
+  )
+
+  test("inner join, one match per row") {
+    withSQLConf(SQLConf.CASE_SENSITIVE.key -> "true") {
+      checkAnswer() { sl =>
+        val upper = upperCaseData(sl)
+        val lower = lowerCaseData(sl)
+        upper.join(lower).where('n === 'N)
+      }
+    }
+  }
+
+  test("inner join, multiple matches") {
+    checkAnswer() { sl =>
+      val x = testData2(sl).where($"a" === 1).as("x")
+      val y = testData2(sl).where($"a" === 1).as("y")
+      x.join(y).where($"x.a" === $"y.a")
+    }
+  }
+
+  test("inner join, no matches") {
+    checkAnswer() { sl =>
+      val x = testData2(sl).where($"a" === 1).as("x")
+      val y = testData2(sl).where($"a" === 2).as("y")
+      x.join(y).where($"x.a" === $"y.a")
+    }
+  }
+
+  ignore("big inner join, 4 matches per row") {
```

Comment:
Why is this ignored?

Reply:
Casting error :( 
`java.lang.ClassCastException: org.apache.spark.sql.execution.UnionExec cannot be cast to edu.berkeley.cs.rise.opaque.execution.OpaqueOperatorExec`
Should be a supported query, but looks we need to fix some bugs.